### DISCUSSION TITLE
Custom controls, multi-select, and replacing points

### DISCUSSION THREAD

**Discussion by DinosaurDad** [fabricjs/fabric.js#7662] [Q&A]
Hi, I recently upgraded a 3.6.x project to 4.6.0, and now 5.0.0. The project uses custom controls, and the way I implemented controls with 3.6, they were beyond broken with 4.x and up. So I decided to rethink/rewrite the controls, and largely based the effort on your polygon editor sample. I have included a sample in this code sandbox:

https://codesandbox.io/s/recursing-robinson-mo15r?file=/index.html

The arrow is the one with custom controls. You can drag the ends, and the script recomputes the points for the arrow. It works really well, without any scaling or translation gotchas. You can also drag the arrow, and the line ends continue to work fine.

The issue comes in when I select both the circle and the arrow, and move them together. The arrow goes where it's supposed to,  and when selected individually again, the controls show up in the right place. But as soon as I drag one of the controls, the arrow appears to remember where it was prior to the multi-select drag.

I'm not sure how to clean up this one remaining issue. Any help would be much appreciated.

https://user-images.githubusercontent.com/523433/153079521-14cf76a6-ebaa-4ebd-888a-cb54521e8025.mp4

**Comment by asturur** [comment]
I need to set time aside to look at this, i m sure we can fix

**Reply by DinosaurDad** [reply]
I've still not gotten anywhere with this. If you get a chance to take a peek I would appreciate it.

**Comment by ShaMan123** [comment]
~~I am guessing it's group (subclassed by ActiveSelection) changing coords of object and active selection getting destroyed on deselect without restoring them~~
There's another open issue resembling this #7660
~~Group is phasing out and there's an open PR of the new group, could you test it with the new Group?~~

**Reply by ShaMan123** [reply]
Well I tried and I managed to reproduce with the new Group so it might be an error in design of the new Group as well (relative planes are tough) 
But I don't think so.

**Reply by ShaMan123** [reply]
I even got the arrow to scale, not only to translate by itself

**Reply by ShaMan123** [reply]
I have managed to make it worse https://codesandbox.io/s/hls83m
seems like the point calculation doesn't take into account transformation

**Reply by ShaMan123** [reply]
I battled something similar in group.
Are points relative to polygon? If so it can explain this mess.
You might need to account for the change of the center of the object or something.
It's easy to get this part mixed up between relating to polygon or the canvas

**Comment by ShaMan123** [comment]
related somehow #7639?

**Reply by DinosaurDad** [reply]
My confusion comes in because when it's not in a selection group, everything works as expected.

**Reply by ShaMan123** [reply]
it is really confusing

**Reply by ShaMan123** [reply]
Maybe try to triage by making it a part of Group (adding/removing it) and seeing what happens. Regardless of selection.
And I suggest you make sure it's visual state correlates to it's actual state (it might be dirty be not repaint until we see something strange happening but the bug might be before)

**Comment by DinosaurDad** [comment]
Maybe I can ask this another way. The `shape.on('modified', ...)` call is no longer fired in a multi-select in FabricJS 4.x, but it used to in FabricJS 3.6. Is there an easy way to add that functionality back into my code?

I want to fire a modified handler for each object in the selection after drag is finished.

**Comment by DinosaurDad** [comment]
What I'm trying to accomplish still has not been sorted out. I either need to figure out how to prevent these objects from participating in multi-select (not nice), keep the objects from jumping after movement in a multi-select, or properly fire the object:modified handler the way 3.6 used to. 

Anyone available for paid resolution of the jumping line ends so multi-select can still be used?

If not, how can I prevent an object from becoming part of a multi-select, but available for single-select?

**Comment by ShaMan123** [comment]
Well first of all #7940 
Second I think that if you have tested it with the new group #7670 then it means your logic is inadequate.
By the looks of it (a superficial look) the points you are extracting seem to be wrong. And should at least be updated during `setCoords`. I dare saw they should be the coords themselves.
If anything, I suggest keeping in mind to which coordinate plane you are relating to - that is how I fixed Group. It is a major bug that lies under the radar. And seems to be the case here. 
The points AFAIK relate to the canvas plane which is absolutely WRONG.

**Reply by DinosaurDad** [reply]
Thanks for the reply. I'm confused. I'm not using a group at all. When the object is manipulated directly (selection move, or control move) it works great. How are groups involved? I'm not sure what coordinates I need to update differently. None of my code gets called when the multi-select is moved.

It doesn't make sense that individually it works completely differently than when it's multi-selected. Events aren't fired the same, and everything looks good visually.

The root of the problem for me is that I don't know what is at play here, and your description doesn't make sense to me as a fabric consumer, rather than code maintainer.

Do you guys offer any paid support for this component? It's just a time sink for me at this point.

And the line component referenced in 7940 is a hot mess.

**Reply by ShaMan123** [reply]
I'll explain. ActiveSelection is a Group. Once objects are selected they are grouped by an ActiveSelection, what makes them relate to the center of the ActiveSelection as their (0, 0) point. 

> None of my code gets called when the multi-select is moved.

This is the exact point. I suggest you design your code differently. Or hook onto the active selection by extending it (e.g. listen to it's modified event to recalculate the line's points).

Think of yourself standing in the street. Let's say we want to understand where your hand is so we do some measurements. Now someone picks you up in a car. Meanwhile we want to keep tracking the position of your hand. Intuitively it is clear that we should measure the position of your hand in respect to your body or center. It makes no sense to measure your hand's position in respect to where you were standing in the street, even more so to the street's beginning (whatever that is). 
Not sure this helps but it is in fact a usage of `frame of reference`. 
If we measure your hand in respect to your body we don't need to remeasure it when your driving around. We do need to measure the car's position and your position in respect to the car and sum that up.
That is effectively what is happening if I am not mistaken. The canvas or the canvas coordinate plane being the street, ActiveSelection  being the car, you starring as the object and your hand the control points.

Hope that was clear.

> Do you guys offer any paid support for this component?

I do. My email is public on my profile.

**Reply by ShaMan123** [reply]
The polygon example seems to work with ActiveSelection 

https://codepen.io/shaman123/pen/mdXRgYP

**Comment by ShaMan123** [comment]
@asturur I think we should consider integrating the editable polygon controls into the source code. It seems the community is asking for it. Many issues are opened regarding it.

**Comment by DinosaurDad** [comment]
I really appreciate the explanation. I am going to try to reincorporate the polygon sample in my code.
